src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,510 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
+                if (!CanGenerateSwitchDispatch(node, loweredNodes))
                     return false;
+
+                var input = ((BoundTestDecisionDagNode)node).Test.Input;
+                ValueDispatchNode n = GatherValueDispatchNodes(node, loweredNodes, input);
+                LowerValueDispatchNode(n, input);
+                return true;
+            }
+
+            private bool CanGenerateSwitchDispatch(BoundDecisionDagNode node, HashSet<BoundDecisionDagNode> loweredNodes)
+            {
+                switch (node)
+                {
+                    // These are the forms worth optimizing.
+                    case BoundTestDecisionDagNode { WhenFalse: BoundTestDecisionDagNode test2 } test1:
+                        return CanDispatch(test1, test2);
+                    case BoundTestDecisionDagNode { WhenTrue: BoundTestDecisionDagNode test2 } test1:
+                        return CanDispatch(test1, test2);
+                    default:
+                        // Other cases are just as well done with a single test.
+                        return false;
+                }
+
+                bool CanDispatch(BoundTestDecisionDagNode test1, BoundTestDecisionDagNode test2)
+                {
+                    if (loweredNodes.Contains(test2))
+                    {
+                        Debug.Assert(this._dagNodeLabels.ContainsKey(test2));
+                        return false;
+                    }
+                    if (this._dagNodeLabels.ContainsKey(test2))
+                        return false;
+                    var t1 = test1.Test;
+                    var t2 = test2.Test;
+                    if (!(t1 is BoundDagValueTest || t1 is BoundDagRelationalTest))
+                        return false;
+                    if (!(t2 is BoundDagValueTest || t2 is BoundDagRelationalTest))
+                        return false;
+                    if (!t1.Input.Equals(t2.Input))
+                        return false;
+                    return true;
+                }
+            }
+
+            private ValueDispatchNode GatherValueDispatchNodes(
+                BoundDecisionDagNode node,
+                HashSet<BoundDecisionDagNode> loweredNodes,
+                BoundDagTemp input)
+            {
+                IValueSetFactory fac = ValueSetFactory.ForSpecialType(input.Type.SpecialType);
+                return GatherValueDispatchNodes(node, loweredNodes, input, fac);
+            }
+
+            private ValueDispatchNode GatherValueDispatchNodes(
+                BoundDecisionDagNode node,
+                HashSet<BoundDecisionDagNode> loweredNodes,
+                BoundDagTemp input,
+                IValueSetFactory fac)
+            {
+                if (loweredNodes.Contains(node))
+                {
+                    bool foundLabel = this._dagNodeLabels.TryGetValue(node, out LabelSymbol label);
+                    Debug.Assert(foundLabel);
```

Comment:
>Debug.Assert(foundLabel); [](start = 20, length = 25)

Is there a chance we'll get here with `!foundLabel` in recursive call when `WhenTrue` or `WhenFalse` is not a `BoundTestDecisionDagNode`?

Reply:
No.  If there were more than one path to a node it would have been assigned a label. Moreover, the decision dag is acyclic.

---
In reply to: [390511023](https://github.com/dotnet/roslyn/pull/42313#discussion_r390511023) [](ancestors = 390511023)